File: cpp/783_二叉搜索树节点最小距离_easy.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int minDiffInBST(TreeNode* root) {
        int pre = -1;
        int ans = INT_MAX;
        dfs(root, pre, ans);
        return ans;
    }
    
    void dfs(TreeNode *root, int &pre, int &ans) {
        if (!root) return;
        dfs(root->left, pre, ans);
        if (pre != -1) ans = min(ans, root->val - pre);
        cout<<ans<<endl;
        pre = root->val;
        dfs(root->right, pre, ans);
    }
};

File: python/783_二叉搜索树节点最小距离_easy.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def minDiffInBST(self, root: TreeNode) -> int:
        self.pre, self.ans = -1, float('+inf')
        def DFS(root):
            if not root:
                return
            DFS(root.left)
            if self.pre != -1:
                self.ans = min(self.ans, root.val-self.pre)
            self.pre = root.val
            DFS(root.right)
        DFS(root)
        return self.ans